scripts/copy-tables.py

```diff
@@ -0,0 +1,207 @@
+
+    """
+    print(f"...checking table statement: {table}")
+    to_search = re.search("TO(.[.\w]*)", curr_create_table_statement)
+    from_search = re.search("FROM(.[.\w]*)", curr_create_table_statement)
+
+    # make sure we ditch the "default." before checking against SHOW TABLES
+    _, to_local_table = to_search.group(1).strip().split(".")
+    _, from_local_table = from_search.group(1).strip().split(".")
+
+    print(f"...looking for local tables: {to_local_table}, {from_local_table} ")
+
+    all_tables = [result[0] for result in target_client.execute("SHOW TABLES")]
+
+    assert (
+        to_local_table in all_tables
+    ), f"{to_local_table} needs to be created before {table}"
+    assert (
+        from_local_table in all_tables
+    ), f"{from_local_table} needs to be created before {table}"
+    print("...local tables found, mv check complete !\n")
+
+
+def copy_tables(
+    source_client: Client,
+    target_client: Client,
+    database: str,
+    execute: bool,
+    tables: Optional[Sequence[str]],
+) -> None:
+    """
+    When adding a replica to a clickhouse cluster, that node will not have any tables
+    created yet. In order to get the up to date table statements, we need to copy the
+    create table statements and run them on the new node(s).
+
+    The source_client is the clickhouse node that you are copying the table statements
+    from. The target_client is the clickhouse node is the node you want to create the
+    tables on.
+
+    The table engines that have some extra verification are materialized views, and
+    any of the replicated tables that are listed in the clickhouse docs:
+    https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/replication/
+
+    By default running the script will only print out the CREATE TABLE statement, to
+    actually have the CREATE TABLE statement run, use --execute.
+
+    Consider doing one table at a time if you have tables that will take a while to
+    replicate and might increase the cluster's normalized load considerably.
+    """
+
+    if not tables:
+        tables = [result[0] for result in source_client.execute("SHOW TABLES")]
+
+    show_table_statments = OrderedDict()
+
+    for name in tables:
+        ((engine,),) = source_client.execute(
+            f"SELECT engine FROM system.tables WHERE name = '{name}'"
+        )
+
+        ((curr_create_table_statement,),) = source_client.execute(
+            f"SHOW CREATE TABLE {database}.{name}"
+        )
+
+        if engine == "MaterializedView":
+            print("\nMaterialized View Check:")
+            verify_local_tables_exist_from_mv(
+                target_client, curr_create_table_statement, name
+            )
+
+        if engine.startswith("Replicated"):
+            print("\nReplicated Table Check:")
+            verify_zk_replica_path(source_client, curr_create_table_statement, name)
+
+        show_table_statments[name] = curr_create_table_statement
```

Comment:
Is there some sort of ordering guarantees in the way create table commands are run if no tables are provided in the api call? Lets take an example of the materialized view. The sequencing is both (raw_table, aggregated_table) and then materialized_view. Does running this loop order that guarantee?

If not, you could build 2 dictionaries. One with the base tables and another with the materialized_views. And then execute all base table ones before the materialized views one.